core/controllers/feedback.py

```diff
@@ -24,8 +24,18 @@
 from core.domain import suggestion_services
 from core.domain import user_services
 
+from typing import Any, Dict, List, Optional, Tuple, TypedDict, TypeVar
 
-def replace_user_id_with_username_in_dict(thread_dicts, user_keys):
+
+# Here we use type Any because we want to allow list of different
+# TypedDict dictionaries through this type variable.
+ThreadDictType = TypeVar('ThreadDictType', bound=List[Any])
```

Comment:
This seems off to me. Can you make it more specific or justify why we need this level of generality? I don't think we have different kinds of thread dicts and even if we did you should be using an explicit Union rather than Any.

Reply:
If we define a Union of typeddicts here then 3 errors will occur:
1.  https://github.com/oppia/oppia/blob/934e23416ebb513935c124ed2be93e26a2632785/core/controllers/feedback.py#L42 here `username_key` is a sting variable and typedDict with string variable is prohibited ( only string literals are allowed ).
2. https://github.com/oppia/oppia/blob/934e23416ebb513935c124ed2be93e26a2632785/core/controllers/feedback.py#L44 similar error as 1 will occur.
3. https://github.com/oppia/oppia/blob/934e23416ebb513935c124ed2be93e26a2632785/core/controllers/feedback.py#L42 index assignment error, since it is a Union type so mypy don't know which typeddict is assigned/changed.

Reply:
There are only 3 possibilities we can do here:
1. We can use Union of typedDicts as mentioned in above comment.
2. we can use `ThreadDictType = TypeVar('ThreadDictType', bound=List[Mapping[str, Any]])`, but this will again throw an error. Because below we are doing `thread_dict.pop` and the pop method is not defined on mapping dicts.
3. We can use `ThreadDictType = TypeVar('ThreadDictType', bound=List[Union of typedDicts])` but this will throw the same error as 1. So, kind of we has only 2 approaches.

If you have any other approach I'm happy to try.

Reply:
Just for clarity -- since I think this will help with future discussions -- I generally can't parse arguments along the lines of "here are these three things I tried and none of them work". It skips a very important step, which is explaining what the actual issue is. Without understanding the actual issue (which in this case is "what exactly is being represented by each type"), I can't reason through the logic and we need to do more back-and-forth, which makes the discussions longer.

For this case, the core thing I need to understand as a reviewer is: what are the different types of thread dicts that makes this sort of construct necessary? Can we split ThreadDictType into different types that distinguish the meaning of each subtype? (This is a bit like what we did with the to_frontend_dict for stats -- if you are running into contortions with the types then that is likely a sign that you need to define more specific types because you are probably inappropriately treating two different things similarly.)